app/controllers/settings_controller.rb

```diff
@@ -55,6 +55,11 @@ def using_incoming_sms_token_message
   end
 
   private
+    def authorize!(*args)
+      super
+      require_recent_login
+    end
```

Comment:
I did this one the way I did because `load_and_authorize_resource` is not being called. I guess there wouldn't be any danger of checking for a recent login before `authorize!` is called (either by `load_and_authorize_resource` for `UsersController` and `UserBatchController` or directly as in `SettingsController`), but I've been thinking of this as a three-part check: 1) Authentication, 2) Authorization, 3) Stale session check.

This probably gets back to the fact that I originally tried to implement this by adding extra checks at the end of `UserAbility.initialize`. The problem with that approach was that it made `can?` start returning `false` for stale logins, so the UI elements to perform sensitive actions weren't even being rendered.

I then started down the path of overrideing `authorize!` in `ApplicationController` similar to what I've done here in `SettingsController`, but I abandoned that because I thought that having `load_and_authorize` do more magical stuff might be confusing. An explicit `before_action` for checking stale logins seemed easier to maintain.

As for returning a 4XX status code in the `xhr?` case, that sounds reasonable. I'd probably return a 401 instead of a 403 since this is a type of authorization failure, but either one is fine probably.